tests/Domain/DonationServiceTest.php

```diff
@@ -88,6 +70,71 @@ public function testIdentifiesCharityLackingCapabilities(): void
 
         $this->expectException(CharityAccountLacksNeededCapaiblities::class);
 
-        $this->sut->createDonation($donationCreate, $customerId);
+        $this->sut->createDonation($donationCreate, self::CUSTOMER_ID);
+    }
+
+    public function testInitialPersistRunsOutOfRetries(): void
+    {
+        $logger = $this->prophesize(LoggerInterface::class);
+        $logger->info('Donation Create persist before stripe work error: . Retrying 1 of 3.')->shouldBeCalledOnce();
+        $logger->info(Argument::type('string'))->shouldBeCalled();
+        $logger->error('Donation Create persist before stripe work error: . Giving up after 3 retries.')
```

Comment:
Potentially a not-worthwhile amount of work to make the exception more realistic and still of this type?

Trying to do it the normal Prophecy way gives
> Can not add prophecy for a method `Double\Doctrine\DBAL\Exception\UniqueConstraintViolationException\P37::getMessage()`
as it is a final method.

Reply:
Yeah it's awkward because the exception constructor is internal and its final. Maybe that's a hint we should be catching something more general - not sure. But if you don't mind the psalm errors this way gives us a test that checks we're passing the message through:

```diff
--- a/tests/Domain/DonationServiceTest.php
+++ b/tests/Domain/DonationServiceTest.php
@@ -2,6 +2,7 @@
 
 namespace MatchBot\Tests\Domain;
 
+use Doctrine\DBAL\Driver\PDO\Exception;
 use Doctrine\DBAL\Exception\UniqueConstraintViolationException;
 use MatchBot\Application\HttpModels\DonationCreate;
 use MatchBot\Application\Matching\Adapter;
@@ -76,9 +77,9 @@ class DonationServiceTest extends TestCase
     public function testInitialPersistRunsOutOfRetries(): void
     {
         $logger = $this->prophesize(LoggerInterface::class);
-        $logger->info('Donation Create persist before stripe work error: . Retrying 1 of 3.')->shouldBeCalledOnce();
+        $logger->info('Donation Create persist before stripe work error: An exception occurred in the driver: EXCEPTION_MESSAGE. Retrying 1 of 3.')->shouldBeCalledOnce();
         $logger->info(Argument::type('string'))->shouldBeCalled();
-        $logger->error('Donation Create persist before stripe work error: . Giving up after 3 retries.')
+        $logger->error('Donation Create persist before stripe work error: An exception occurred in the driver: EXCEPTION_MESSAGE. Giving up after 3 retries.')
             ->shouldBeCalledOnce();
 
         $this->sut = $this->getDonationService(withAlwaysCrashingEntityManager: true, logger: $logger->reveal());
@@ -101,8 +102,9 @@ class DonationServiceTest extends TestCase
     ): DonationService {
         $emProphecy = $this->prophesize(RetrySafeEntityManager::class);
         if ($withAlwaysCrashingEntityManager) {
-            $exception = $this->prophesize(UniqueConstraintViolationException::class);
-            $emProphecy->persistWithoutRetries(Argument::type(Donation::class))->willThrow($exception->reveal());
+            $emProphecy->persistWithoutRetries(Argument::type(Donation::class))->willThrow(
+                new UniqueConstraintViolationException(new Exception('EXCEPTION_MESSAGE'), null)
+            );
         }
 
```

I need to eat a bit, lets have a call later and get this merged together?

Also for the logger I feel like an in-memory logger would make the test quite a bit simpler to read than mocking the calls to it? Can easily write our own or use one off the shelf like  https://github.com/MetaSyntactical/inmemory-logger/blob/2.2.x/src/InMemoryLogger.php  (I don't remember where I found that one now, no idea if it's the best)

Reply:
I'll be offline before the end of your current session I think @bdsl but suppressing a psalm error sounds OK in this case. Integrating a different test logger seems like it would make sense to do as a later enhancement to me.